### DISCUSSION TITLE
Implement run code lenses as a run, not debug

### DISCUSSION THREAD

**Discussion by kpodsiad** [scalameta/metals#3853] [General]
There are [run/test environment requests](https://build-server-protocol.github.io/docs/extensions/jvm.html#test-environment-request) in BSP which provides all necessary data needed to run main class/test class as a process. I was thinking about using it to reimplement Metals run lenses. Thanks to that change running will be easier than currently and users will be able to provide their input through the console.

In terms of implementation, I was thinking about adding a new provider, namely `terminalCommandRunnerProvider: boolean`. This provider will indicate if the client supports a new client command named, for example, `RunInTerminal` which accepts `string` that needs to be run by the client.

@ckipp01 What does the second part look like from nvim perspective?

**Comment by ckipp01** [comment]
I _think_ that this should be fine. Just to be sure I understand, this would be the flow for this right?

```mermaid
sequenceDiagram
    rect rgb(191, 223, 255)
    note right of client: new - needs to be set by the client
    client->>+client: terminalCommandRunnerProvider: true
    end
    rect rgb(200, 150, 255)
    note right of client: same nothing changes from user perspective
    metals->>+client: textDocument/codeLens (with run command)
    client->>+client: user triggers the run
    client->>+metals: command sent to metals
    end
    rect rgb(191, 223, 255)
    note right of client: new user interaction
    metals->>+client:  RunInTerminal client command sent
    client->>+client: gets user input
    client->>+metals: input sent back
    end
    rect rgb(200, 150, 255)
    note right of client: same nothing changes from user perspective
    client->metals: rest of the process takes over
    end
```
